### PR TITLE
Add debug_one method to all FMRegressors

### PR DESCRIPTION
**Pull request by nakamasato**
# why

To make it easy to check the contribution of each feature and each combination of features in an FM regressor.

# what

1. [Refactor] Introduced common methods to calculate interactions
	- `_calculate_interactions` is moved to `base`, which is calculated by:
		- `_interaction_combination_keys`: combinations of interaction features (e.g. `sigma i < j` for normal FM)
		- `_interaction_val`: value corresponding to a given combination of interaction features (e.g. `xi * xj` for normal FM)
		- `_interaction_coefficient`: coefficient corresponding to a given combination of interaction features (e.g. `latents[xi] * latents[xj]` for normal FM)
1. Added `debug_one` and corresponding `doctest` to the following regressors:
	1. Add `debug_one` in `base` using the common methds introduced above.
	2. Added `_interaction_names`, which is used for feature interaction name in the table created by `debug_one` in `FMRegressor`, `FFMRegressor`, `FwFMRegressor`, and `HOFMRegressor`.

# Related PR

https://github.com/online-ml/river/pull/784

### CODE REVIEW

**Review by MaxHalford — COMMENTED** [review]
Thanks for doing this. Do you think there's some way to refactor the code so it doesn't stutter as much? There's a lot of duplication here

**Comment by nakamasato** [comment]
> Thanks for doing this. Do you think there's some way to refactor the code so it doesn't stutter as much? There's a lot of duplication here

That's true! Let me take a deeper look into it and I'll share if any idea comes up!

**Comment by nakamasato** [comment]
About the interaction part, I think we can generalize by the following three parts:
1. **combination_key**: combinations of feature pairs (e.g. `[(0,1), (1, 2), ...]`) (iterable) 
1. **combination_val**: the corresponding value to the combination keys (e.g. `[x[0]*x[1], x[1]*x[2], ...]`)
1. **coefficient** (or weight): the coefficient corresponding to the combination keys/values (e.g. `[np.dot(latents[0], latents[1]),...]`)

I'm planning to make a private method for each of them and we can reuse them to
- calculate interactions (`_calculate_interactions`) by taking the sum of the **combination_val** multiplied by **coefficient**
- debug (`debug_one`) by getting feature names from **combination_key**, values from **combination_val**, weight from **coefficient**, and contribution from **combination_val** and **coefficient**

For each of the target algorithms, I consider in the following way:

1. FM

    - combination_key
        ```python
        itertools.combinations(x.keys(), 2)
        ```
    - combination_val
        ```python
        functools.reduce(
            lambda x, y: x * y, (x[j] for j in combination)
        ) # same as x[i] * x[j]
        ```
    - coefficient
        ```python
        i, j = combination
        latents[i] * latents[j]
        ```

1. FFM
    - combination_key
        ```python
        itertools.combinations(x.keys(), 2)
        ```
    - combination_val
        ```python
        functools.reduce(
            lambda x, y: x * y, (x[j] for j in combination)
        ) # same as x[i] * x[j]
        ```
    - coefficient
        ```python
        i, j = combination
        latents[i][field(j)] * latents[j][field(i)]
        ```

1. HOFM

    - combination_key
        ```python
        (combinations(x.keys(), d) for d in range(2, self.degree+1))
        ```
    - combination_val
        ```python
        functools.reduce(
            lambda x, y: x * y, (x[j] for j in combination)
        )
        ```
    - coefficient
        ```python
        sum(
            functools.reduce(
                lambda x, y: np.multiply(x, y),
                (self.latents[j][d] for j in combination),
            )
        )
        ```

1. FwFM

    - combination_key
        ```python
        itertools.combinations(x.keys(), 2)
        ```
    - combination_val
        ```python
        functools.reduce(
            lambda x, y: x * y, (x[j] for j in combination)
        ) # same as x[i] * x[j]
        ```
    - coefficient
        ```python
        i, j = combination
        w_int[field(i) + field(j)] * latents[i] * latents[j]
        ```

WDYT? If you have any questions or better ideas, please let me know thanks!

**Comment by MaxHalford** [comment]
That looks @nakamasato, you've convinced me! @gbolmier do you have time to take look?

**Comment by MaxHalford** [comment]
@nakamasato, @gbolmier told me he agrees too :)

**Comment by nakamasato** [comment]
@MaxHalford I implemented the shared idea. In detail, it's slightly different but the basic idea is the same! Please take a look at it when you have time, Thank you! I could move `debug_one` to `Base` class as well.

File: river/facto/base.py [hunk 1]
```diff
@@ -122,9 +125,103 @@ def _field(self, j):
...
+        interaction_combination_keys: sigma (i < j)
+        interaction_val: xi * xj
+        interaction_coefficient: latents[xi] * latents[xj]
+        """
+        return sum(
+            [
+                self._interaction_coefficient(combination)
+                * self._interaction_val(x, combination)
+                for combination in self._interaction_combination_keys(x)
+            ]
```
Comment (MaxHalford) [inline river/facto/base.py:140]: You don't need the `[` and `]` here

File: river/facto/base.py [hunk 2]
```diff
@@ -122,9 +125,103 @@ def _field(self, j):
...
+        )
+
+    def debug_one(self, x: dict, decimals: int = 5) -> str:
+        """Debugs the output of the FM regressor.
```
Comment (MaxHalford) [inline river/facto/base.py:144]: Try to add lines to separate parts, like this:

```
"""Debugs the output of the FM regressor.

Parameters
----------
"""

File: river/facto/base.py [hunk 3]
```diff
@@ -122,9 +125,103 @@ def _field(self, j):
...
+        Parameters
+        ----------
+        x
+            A dictionary of features.
+        decimals
+            The number of decimals use for printing each numeric value.
```
Comment (MaxHalford) [inline river/facto/base.py:150]: Same here

**Review by MaxHalford — APPROVED** [review]
Thanks I really like it! I just made a few suggestions.

Could you add an entry to `docs/releases/unreleased.md`?

**Comment by nakamasato** [comment]
@MaxHalford Thank you for your quick feedback! I'll update it tomorrow!

File: river/facto/base.py (hunk 1 above)
Comment (nakamasato) [reply river/facto/base.py]: Fixed in https://github.com/online-ml/river/pull/787/commits/20ae38f372a3aca1a44b8a17b60076b3cb6e9de7

File: river/facto/base.py (hunk 2 above)
Comment (nakamasato) [reply river/facto/base.py:142]: Fixed in https://github.com/online-ml/river/pull/787/commits/20ae38f372a3aca1a44b8a17b60076b3cb6e9de7

File: river/facto/base.py (hunk 3 above)
Comment (nakamasato) [reply river/facto/base.py:149]: Fixed in https://github.com/online-ml/river/pull/787/commits/20ae38f372a3aca1a44b8a17b60076b3cb6e9de7

**Comment by nakamasato** [comment]
@MaxHalford I've fixed it!

File: docs/releases/unreleased.md [hunk 4]
```diff
@@ -1 +1,6 @@
 # Unreleased
+
+## facto
+
+- Moved `_calculate_interactions` from each of FM implementations to `BaseFM`.
+- Addd `debug_one` method to `BaseFM`.
```
Comment (MaxHalford) [inline docs/releases/unreleased.md:6]: `Added`

**Review by MaxHalford — APPROVED** [review]

File: docs/releases/unreleased.md (hunk 4 above)
Comment (nakamasato) [reply docs/releases/unreleased.md]: https://github.com/online-ml/river/pull/787/commits/87a4ccaa3dd917148c8c3a1c9572f84da0031815 fixed!

**Comment by MaxHalford** [comment]
Great job @nakamasato, thanks! It's nice to have you on board.

**Pull request merged**